app/src/main/java/org/thoughtcrime/securesms/conversation/v2/AttachmentDownloadHelper.kt

```diff
@@ -0,0 +1,119 @@
+) {
+    companion object {
+        private const val BUFFER_TIMEOUT_MILLS = 500L
+        private const val BUFFER_MAX_ITEMS = 10
+        private const val LOG_TAG = "AttachmentDownloadHelper"
+    }
+
+    private val downloadRequests = Channel<DatabaseAttachment>(UNLIMITED)
+
+    init {
+        scope.launch(Dispatchers.Default) {
+            downloadRequests
+                .receiveAsFlow()
+                .timedBuffer(BUFFER_TIMEOUT_MILLS, BUFFER_MAX_ITEMS)
+                .map { attachments ->
+                    val pendingAttachmentIDs = storage
+                        .getAllPendingJobs(AttachmentDownloadJob.KEY, AttachmentUploadJob.KEY)
+                        .values
+                        .mapNotNullTo(hashSetOf()) {
+                            (it as? AttachmentUploadJob)?.attachmentID
+                                ?: (it as? AttachmentDownloadJob)?.attachmentID
+                        }
+
+                    val messagesByID = mmsDatabase.getMessages(attachments.map { it.mmsId })
+                        .associateBy { it.id }
+
+                    // Before handling out attachment to the download task, we need to
+                    // check the requisite for that attachment. This check is very likely to be
+                    // performed again in the download task, but adding stuff into job system
+                    // is expensive so we need to avoid spawning new task whenever we can.
+                    attachments.filter { attachment ->
+                        attachment.attachmentId.rowId !in pendingAttachmentIDs &&
+                                eligibleForDownloadTask(attachment, messagesByID[attachment.mmsId])
+                    }
+                }
+                .flatten()
+                .collect { attachment ->
+                    jobQueue.add(
+                        AttachmentDownloadJob(
+                            attachmentID = attachment.attachmentId.rowId,
+                            databaseMessageID = attachment.mmsId
+                        )
+                    )
+                }
+        }
+    }
+
+    private fun eligibleForDownloadTask(attachment: DatabaseAttachment, message: MessageRecord?): Boolean {
+        if (message == null) {
+            Log.w(LOG_TAG, "Message ${attachment.mmsId} not found for attachment ${attachment.attachmentId}")
+            return false
+        }
+
+        assert(message.id == attachment.mmsId) { "Message ID mismatch: ${message.id} != ${attachment.mmsId}" }
+
+        if (message.isOutgoing) return true
+
+        val sender = message.individualRecipient.address.serialize()
+        val contact = storage.getContactWithSessionID(sender)
+
+        if (contact == null) {
+            Log.w(LOG_TAG, "Contact not found")
+            return false
+        }
+
+        val threadRecipient = storage.getRecipientForThread(message.threadId)
+        if (threadRecipient == null) {
```

Comment:
do we have to retrieve a `thread` AND a `Contact` before we can return a `boolean ||`?

Is it not valid if we have a thread, but no contact?